python/dgl/mock_sparse2/reduction.py

```diff
@@ -0,0 +1,322 @@
+"""DGL sparse matrix reduce operators."""
+from typing import Optional
+
+import torch
+
+from .sparse_matrix import SparseMatrix
+
+
+def reduce(A: SparseMatrix, dim: Optional[int] = None, rtype: str = "sum"):
+    """Compute the reduction of non-zero values in sparse matrix A along
+    the given dimension :attr:`dim`.
+
+    If :attr:`dim` is None, it reduces all the elements in the sparse
+    matrix. Otherwise, it reduces on the row (``dim=0``) or column (``dim=1``)
+    dimension, producing a tensor of shape ``(A.shape[1], ) + A.val.shape[:1]``
+    or ``(A.shape[0],) + A.val.shape[:1]``.
+
+    The reduction does not count zero values. If the row or column to be
+    reduced does not have any non-zero value, the result will be 0.
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix
+    dim : int, optional
+        The dimension to reduce.  Must be either 0 (by rows) or 1 (by columns).
+
+        If not given, the reduction will happen across all non-zero entries.
+    rtype: str
+        Reduction type, one of ``['sum', 'smin', 'smax', 'smean', 'prod']``,
```

Comment:
if s to callout the smax is for sparse matrix, which will only operate on non-zero entries, then prod should also be sprod.